File: app/src/main/java/com/xnote/app/domain/model/BackgroundKey.kt
package com.xnote.app.domain.model

// -- Type Definitions

sealed interface BackgroundKey {
    data class Builtin(val id: String) : BackgroundKey {
        init { require(id in BuiltinBackgroundIds) { "Unsupported built-in background: $id" } }
    }

    data class Image(val attachmentId: String, val maskOpacity: Int = 72) : BackgroundKey {
        init {
            require(attachmentId.isNotBlank()) { "Missing background attachment ID" }
            require(maskOpacity in 0..100) { "Background mask opacity must be between 0 and 100" }
        }
    }
}

// -- Constants

const val DefaultBuiltinBackgroundId = "default"
const val CreamBuiltinBackgroundId = "cream"
const val RuledBuiltinBackgroundId = "ruled"
const val GridBuiltinBackgroundId = "grid"

private val BuiltinBackgroundIds = setOf(
    DefaultBuiltinBackgroundId,
    CreamBuiltinBackgroundId,
    RuledBuiltinBackgroundId,
    GridBuiltinBackgroundId,
)

// -- Functions

fun BackgroundKey.encode(): String = when (this) {
    is BackgroundKey.Builtin -> "builtin:$id"
    is BackgroundKey.Image -> "image:$maskOpacity:$attachmentId"
}

fun parseBackgroundKey(raw: String?): BackgroundKey? {
    val id = raw?.substringAfter(':', "") ?: return null
    return when {
        raw.startsWith("builtin:") && id in BuiltinBackgroundIds -> BackgroundKey.Builtin(id)
        raw.startsWith("image:") -> {
            val opacity = id.substringBefore(':').toIntOrNull()
            val attachmentId = id.substringAfter(':', "")
            if (opacity != null && opacity in 0..100 && attachmentId.isNotBlank()) {
                BackgroundKey.Image(attachmentId, opacity)
            } else null
        }
        else -> null
    }
}

fun defaultBackgroundKey(): BackgroundKey = BackgroundKey.Builtin(DefaultBuiltinBackgroundId)

fun resolveBackgroundKey(
    noteBackground: BackgroundKey?,
    defaultBackground: BackgroundKey,
): BackgroundKey = noteBackground ?: defaultBackground

File: flutter/lib/domain/model/background_key.dart
// -- Type Definitions

final class BackgroundKey {
  factory BackgroundKey(String id) {
    if (!builtinBackgroundIds.contains(id)) {
      throw ArgumentError.value(id, 'id', 'Unsupported built-in background');
    }
    return BackgroundKey._(id);
  }

  const BackgroundKey._(this.id);

  final String id;

  // -- Functions

  String encode() => 'builtin:$id';

  @override
  bool operator ==(Object other) => other is BackgroundKey && other.id == id;

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() => 'BackgroundKey($id)';
}

// -- Constants

const defaultBuiltinBackgroundId = 'default';
const creamBuiltinBackgroundId = 'cream';
const ruledBuiltinBackgroundId = 'ruled';
const gridBuiltinBackgroundId = 'grid';
const builtinBackgroundIds = <String>{
  defaultBuiltinBackgroundId,
  creamBuiltinBackgroundId,
  ruledBuiltinBackgroundId,
  gridBuiltinBackgroundId,
};

// -- Functions

BackgroundKey? parseBackgroundKey(String? raw) {
  const prefix = 'builtin:';
  if (raw == null || !raw.startsWith(prefix)) {
    return null;
  }
  final id = raw.substring(prefix.length);
  return builtinBackgroundIds.contains(id) ? BackgroundKey(id) : null;
}

BackgroundKey defaultBackgroundKey() =>
    BackgroundKey(defaultBuiltinBackgroundId);

BackgroundKey resolveBackgroundKey({
  required BackgroundKey? noteBackground,
  required BackgroundKey defaultBackground,
}) =>
    noteBackground ?? defaultBackground;
